server/logging/firehose.go

```diff
@@ -0,0 +1,162 @@
+package logging
+
+import (
+	"encoding/json"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/firehose"
+	"github.com/aws/aws-sdk-go/service/firehose/firehoseiface"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+)
+
+const (
+	firehoseMaxRetries = 8
+
+	// See
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/firehose/#Firehose.PutRecordBatch
+	// for documentation on limits.
+	firehoseMaxRecordsInBatch = 500
+	firehoseMaxSizeOfRecord   = 1000 * 1000     // 1,000 KB
+	firehoseMaxSizeOfBatch    = 4 * 1000 * 1000 // 4 MB
+)
+
+type firehoseLogWriter struct {
+	client firehoseiface.FirehoseAPI
+	stream string
+	logger log.Logger
+}
+
+func NewFirehoseLogWriter(region, id, secret, stream string, logger log.Logger) (*firehoseLogWriter, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(id, secret, ""),
+		Region:      &region,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "create Firehose client")
+	}
+	client := firehose.New(sess)
+
+	f := &firehoseLogWriter{
+		client: client,
+		stream: stream,
+		logger: logger,
+	}
+	if err := f.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Firehose writer")
+	}
+	return f, nil
+}
+
+func (f *firehoseLogWriter) validateStream() error {
+	out, err := f.client.DescribeDeliveryStream(
+		&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: &f.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", f.stream)
+	}
+
+	if (*(*out.DeliveryStreamDescription).DeliveryStreamStatus) != firehose.DeliveryStreamStatusActive {
+		return errors.Errorf("delivery stream %s not active", f.stream)
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) Write(logs []json.RawMessage) error {
+	var records []*firehose.Record
+	totalBytes := 0
+	for _, log := range logs {
+		// We don't really have a good option for what to do with logs
+		// that are too big for Firehose. This behavior is consistent
+		// with osquery's behavior in the Firehose logger plugin, and
+		// the beginning bytes of the log should help the Fleet admin
+		// diagnose the query generating huge results.
+		if len(log) > firehoseMaxSizeOfRecord {
```

Comment:
Just a question. Have you tried inserting a log 1 byte less than the max to a real firehose server and seeing if it makes it across?

I know it _should_, I just would hate for us to hit an edge case where the library is adding some encoding to the logs behind the scenes that actually causes them to exceed this limit.

Reply:
I was able to confirm that I could put 4 records of exactly 1MB each and retrieve 4MB on the other end. Seems that the docs are accurate (https://docs.aws.amazon.com/sdk-for-go/api/service/firehose/#Firehose.PutRecordBatch)

> Each PutRecordBatch request supports up to 500 records. Each record in the request can be as large as 1,000 KB (before 64-bit encoding), up to a limit of 4 MB for the entire request. These limits cannot be changed.